branches/0.9.x/man/cwstudio.1
.TH cwstudio 1  "October 29, 2010" "" ""
.SH NAME
cwstudio \- lightweight and portable CW signals generator
.SH SYNOPSIS
.B cwstudio
[\-\-play]
[\-\-noplay]
[\-\-silent]
[\-\-verbose]
[\-\-groups]
[\-\-words]
[\-\-calls]
[\-\-agc 0-100]
[\-\-bits bits]
[\-\-click 0-100]
[\-\-chars nchars]
[\-\-charset charset]
[\-\-cspaces 0-100]
[\-\-dashlen 100-10000]
[\-\-detune 0-100]
[\-\-even 0-100]
[\-\-freq 50-4000]
[\-\-number 1-100]
[\-\-hand 0-100]
[\-\-highcut 300-10000]
[\-\-hum 0-100]
[\-\-lowcut 50-1000]
[\-\-noise 0-100]
[\-\-odd 0-100]
[\-\-output outfile]
[\-\-qsb 0-100]
[\-\-samplerate samplerate]
[\-\-seed 1-32767]
[\-\-shape -10 - 10]
[\-\-signals 1-5]
[\-\-spacelen 20-300]
[\-\-sweep -4000 - 4000]
[\-\-sweepness 0 - 10000]
[\-\-tempo 5-500]
[\-\-window 0-1000]
[\-\-wordset 0-1000]
[\-\-wspaces 0-100] 
.SH DESCRIPTION
Generates 1-5 mixed CW signals of various properties, adds optional noise and saves it to a WAV file. If compiled on proper environment, it can play the result.
.PP
All text output is written to stderr. If stdout is redirected, the created file is written here. If stdin is redirected, the generated text is taken from it.
.PP
Optional CWPARAM environment variable can be set to hold some parameters, for example CWPARAM="--even 10 --odd 10".
.SH OPTIONS
.TP
\-\-play \-\-noplay
Play the result (default), otherwise only save the WAV file.
.TP
\-\-silent \-\-verbose
Verbose mode gives full output (default), silent mode presents only generated text.
.TP
\-\-groups \-\-words \-\-calls
The "groups" mode generates random five-letter groups. The "words" mode generates most common English words. The "calls" mode generates real amateur callsigns.
.TP
\-\-agc 0-100
Simulate AGC response of receiver by varying noise volume along RMS of the signal. Default is 100.
.TP
\-\-bits bits
Set bitrate. Default is 16.
.TP
\-\-click 0-100
Simulate click by lowering sustain part of tone at given level (in dB) below attack phase. Default is 1 dB.
.TP
\-\-chars nchars
Take first "nchars" from default charset in "groups" mode.
.TP
\-\-charset charset
Set customized charset for random group generation in "groups" mode.
.TP
\-\-cspaces 0-100
Set additional spaces (one space has a length of a dot) between chars. Default is 0.
.TP
\-\-dashlen 100-1000
Length of dash, in percentage of dot. Default is 300% (3:1).
.TP
\-\-detune 0-100
Simulate frequency drift of the signal. Default is 0.
.TP
\-\-even 0-100
Enhance the signal by given percent of even harmonics. Default is 0. Values 0-5 are most valuable in practice.
.TP
\-\-freq 50-4000
Use given frequency in Hz..
.TP
\-\-number 1-100
Generate given number of groups/words/calls. Default is 20.
.TP
\-\-hand 0-100
Simulate hand transmitting by introduce random errors in dash/dot lengths.
.TP
\-\-highcut 300-10000
Set high cutoff frequency (Hz) of generated noise. Default is 2400.
.TP
\-\-hum 0-100
Add given percentage of 50 Hz mains hum. Default is 0.
.TP
\-\-lowcut 50-1000
Set low cutoff frequency (Hz) of generated noise. Default is 300.
.TP
\-\-noise 0-100
Add given percentage of noise. Default is 100.
.TP
\-\-odd odd
Enhance the signal by given percent of odd harmonics. Default is 0. Values 0-5 are most valuable in practice.
.TP
\-\-output outfile
Store the result in a WAV file of given name. Default is output.wav.
.TP
\-\-qsb 0-100
Add given amount of signal QSB. Default is 0.
.TP
\-\-samplerate samplerate
Set given samplerate. Default is 44100.
.TP
\-\-seed 0-32767
Set explicitly the seed of random generator. If not set or set to 0, seed is also random generated. Otherwise it is set to given value, resulting in a possibility of repetitive random content generation.
.TP
\-\-shape -50 - +50
Set the shape of random generated numbers. Positive value increases the fraction of higher numbers (last letters, words, calls). Negative values prefer first ones. Zero means flat uniform distribution. 
Can be used to increase the number of newly learned letters/words. In practice, values -10 to 10 are well working.
.TP
\-\-signals 1-5
Generate given number of mixed signals, each of hardcoded frequency ratio, tempo ratio and amplitude ratio. Every signal plays the same text in a loop. Default is 3.
.TP
\-\-spacelen 20-300
Space length (between dots and dashes) in percentage of dot. Default is 100% (1:1).
.TP
\-\-sweep -4000 - 4000
Simulate sweep (chirp or filter ringing according to parameters) by starting each tone from given frequency (can be negative, which results in zero crossing). Frequency is then changed exponentially to this given in --freq with a rate given by --sweepness. Default is 0.
.TP
\-\-sweepness 0-10000
The rate of exponential frequency change from --sweep to --freq. Default is 0 (no sweep).
.TP
\-\-tempo 5-500
Tempo of generated main signal in cpm (chars per minute) according to PARIS group.
.TP
\-\-window 0-1000
Raised cosine window width (used to avoid clicks in each tone). Default is 100 samples.
.TP
\-\-wordset 0-1000
Take only first given number of most common English words in "words" mode.
.TP
\-\-wspaces 0-100
Set additional spaces between words/groups/calls. Default is zero.
.PP
.SH EXIT STATUS
CWStudio returns 0 if everything is OK. In the case of error, return value is different.
.SH AUTHOR
Lukasz Komsta, SP8QED, 2008-2011


ChangeLog
2011-10-31 : Fixed a --qsb bug (reverted "if" condition)

2011-10-30 : Fixed a bug in --silent option.

2011-10-29 : RELEASE 0.9.0


src/cwstudio.c
/*$T src/cwstudio.c GC 1.140 11/05/11 20:22:22 */

/*$I0

    This file is part of CWStudio.

    Copyright 2008-2011 Lukasz Komsta, SP8QED

    CWStudio is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CWStudio is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CWStudio. If not, see <http://www.gnu.org/licenses/>.

 */
#include "cwgen.h"
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#ifdef HAVE_CTYPE_H
#include <ctype.h>
#endif
#define RANGE(x, l, u) \
	if(param.x < l) param.x = l; \
	if(param.x > u) param.x = u;
#if HAVE_WINDOWS_H
#include <windows.h>
#endif
#ifdef HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#ifdef HAVE_SYS_SOUNDCARD_H
#include <sys/soundcard.h>
#elif defined HAVE_SOUNDCARD_H
#include <soundcard.h>
#elif defined HAVE_MACHINE_SOUNDCARD_H
#include <machine/soundcard.h>
#endif

/*
 =======================================================================================================================
    Play audio sample, using WMM or OSS
 =======================================================================================================================
 */
void playsample(cw_sample *sample)
{
	/*~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
#ifdef HAVE_LIBWINMM
	HWAVEOUT		h;
	WAVEFORMATEX	wf;
	WAVEHDR			wh;
	HANDLE			d;
#elif defined HAVE_OSS
	int				audio;
	int				format, stereo;
	int				speed;
#endif
	/*~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#ifdef HAVE_LIBWINMM
	wf.wFormatTag = WAVE_FORMAT_PCM;
	wf.nChannels = 1;
	wf.wBitsPerSample = sample->bits;
	wf.nSamplesPerSec = sample->samplerate;
	wf.nBlockAlign = wf.nChannels * wf.wBitsPerSample / 8;
	wf.nAvgBytesPerSec = wf.nSamplesPerSec * wf.nBlockAlign;
	wf.cbSize = 0;
	d = CreateEvent(0, FALSE, FALSE, 0);
	if(waveOutOpen(&h, 0, &wf, (DWORD) d, 0, CALLBACK_EVENT) != MMSYSERR_NOERROR);
	wh.lpData = sample->data;
	wh.dwBufferLength = (sample->bits / 8) * sample->length - 2;
	wh.dwFlags = 0;
	wh.dwLoops = 0;
	if(waveOutPrepareHeader(h, &wh, sizeof(wh)) != MMSYSERR_NOERROR);
	ResetEvent(d);
	if(waveOutWrite(h, &wh, sizeof(wh)) != MMSYSERR_NOERROR);
	if(WaitForSingleObject(d, INFINITE) != WAIT_OBJECT_0);
	if(waveOutUnprepareHeader(h, &wh, sizeof(wh)) != MMSYSERR_NOERROR);
	if(waveOutClose(h) != MMSYSERR_NOERROR);
	CloseHandle(d);
#elif defined HAVE_OSS
	if((audio = open("/dev/dsp", O_WRONLY, 0)) == -1);
	if((sample->bits == 8))
		format = AFMT_U8;
	else
		format = AFMT_S16_LE;
	if(ioctl(audio, SNDCTL_DSP_SETFMT, &format) == -1);
	stereo = 0;
	if(ioctl(audio, SNDCTL_DSP_STEREO, &stereo) == -1);
	speed = sample->samplerate;
	if(ioctl(audio, SNDCTL_DSP_SPEED, &speed) == -1);
	if(write(audio, sample->data, (sample->bits / 8) * sample->length - 2) == -1);
	if(close(audio) == -1);
#endif
}

/*
 =======================================================================================================================
    Separate arguments form environmental variable to create argv-like table. Function based on "C FAQ" book by Steve
    Summit.
 =======================================================================================================================
 */
int separg(char *options, char *argv[], int size)
{
	/*~~~~~~~~~~~~~~~~~~~*/
	char	*str = options;
	int		i, argc = 0;
	/*~~~~~~~~~~~~~~~~~~~*/

	for(i = 0; i < size; i++) {
		while(isspace((int) *str)) str++;
		if(*str != '\0')
			argv[argc++] = str;
		else {
			argv[argc] = 0;
			break;
		}

		while((*str != '\0') && !isspace((int) *str)) str++;
		if((*str != '\0') && i < (size - 1)) *str++ = '\0';
	}

	return(argc);
}

/*
 =======================================================================================================================
    CWStudio - command line interface
 =======================================================================================================================
 */
int main(int argc, char **argv)
{
	/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
	char				*text = NULL, *morsetext;
	cw_sample			asound, csound;
	cw_param			param;
	int					i, c, val, err;
	static int			play = 1, output = 1, mode = 0, wordset = 100, chars;
	static unsigned int bits = 16;
	static unsigned int samplerate = 44100;
	static char			filename[256] = "output.wav";
	static char			charset[256] = "abstgjnokqfmzixdrhewlypvcu8219376450?!/=";
	char				*envv[20], *envs, *totalv[256];
	int					envc = 0;
	/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

	/* Initialize parameters */
	cw_initparam(&param);

	/* If environmental variable is set, parse it */
	if((envs = getenv("CWPARAM")) != NULL) {
		envc = separg(envs, envv, 20);
		if(envc > 0) {
			totalv[0] = argv[0];
			for(i = 0; i < envc; i++) totalv[i + 1] = envv[i];
			for(i = envc; i < (envc + argc); i++) totalv[i + 1] = argv[i - envc + 1];
		}
	}

	/* else copy argv only */
	else
		for(i = 0; i < argc; i++) totalv[i] = argv[i];

	/* Parse command line parameters */
	while(1) {

		/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
		int						option_index = 0;
		static struct option	long_options[] =
		{
			{ "play", no_argument, &play, 1 },
			{ "noplay", no_argument, &play, 0 },
			{ "silent", no_argument, &output, 0 },
			{ "verbose", no_argument, &output, 1 },
			{ "groups", no_argument, &mode, 0 },
			{ "words", no_argument, &mode, 1 },
			{ "calls", no_argument, &mode, 2 },
			{ "agc", required_argument, NULL, 'a' },
			{ "bits", required_argument, NULL, 'b' },
			{ "click", required_argument, NULL, 'l' },
			{ "chars", required_argument, NULL, 'c' },
			{ "charset", required_argument, NULL, 'C' },
			{ "cspaces", required_argument, NULL, 's' },
			{ "dashlen", required_argument, NULL, 'D' },
			{ "detune", required_argument, NULL, 'd' },
			{ "even", required_argument, NULL, 'E' },
			{ "freq", required_argument, NULL, 'f' },
			{ "number", required_argument, NULL, 'n' },
			{ "hand", required_argument, NULL, 'H' },
			{ "highcut", required_argument, NULL, 'y' },
			{ "hum", required_argument, NULL, 'h' },
			{ "lowcut", required_argument, NULL, 'x' },
			{ "noise", required_argument, NULL, 'N' },
			{ "odd", required_argument, NULL, 'O' },
			{ "output", required_argument, NULL, 'o' },
			{ "qsb", required_argument, NULL, 'q' },
			{ "samplerate", required_argument, NULL, 'B' },
			{ "seed", required_argument, NULL, 'r' },
			{ "shape", required_argument, NULL, 'X' },
			{ "signals", required_argument, NULL, 'S' },
			{ "spacelen", required_argument, NULL, 'L' },
			{ "sweep", required_argument, NULL, 'p' },
			{ "sweepness", required_argument, NULL, 'P' },
			{ "tempo", required_argument, NULL, 't' },
			{ "window", required_argument, NULL, 'w' },
			{ "wordset", required_argument, NULL, 'v' },
			{ "wspaces", required_argument, NULL, 'W' },
			{ 0, 0, 0, 0 }
		};
		/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

		c = getopt_long_only
			(
				argc + envc,
				totalv,
				"a:b:B:c:C:d:D:E:f:g:h:H:l:L:n:o:O:p:P:q:r:s:S:t:v:w:W:x:X:y:",
				long_options,
				&option_index
			);
		if(c == -1) break;
		switch(c)
		{
		case 'a':
			param.agc = atoi(optarg);
			RANGE(agc, 0, 100);
			break;

		case 'b':
			bits = atoi(optarg);
			if((bits != 8) && (bits != 16)) bits = 16;
			break;

		case 'B':
			samplerate = atoi(optarg);
			if
			(
				(samplerate != 8000)
			&&	(samplerate != 11025)
			&&	(samplerate != 22050)
			&&	(samplerate != 44100)
			&&	(samplerate != 16000)
			&&	(samplerate != 24000)
			&&	(samplerate != 48000)
			&&	(samplerate != 96000)
			&&	(samplerate != 192000)
			) samplerate = 44100;
			break;

		case 'c':
			chars = atoi(optarg);
			if((chars > 0) && (chars < 256)) charset[chars] = '\0';
			break;

		case 'C':
			strncpy(charset, optarg, 256);
			break;

		case 'd':
			param.detune = atoi(optarg);
			RANGE(detune, 0, 100);
			break;

		case 'D':
			param.dashlen = atoi(optarg);
			if(param.dashlen == 0) param.dashlen = 300;
			RANGE(dashlen, 100, 10000);
			break;

		case 'E':
			param.even = atoi(optarg);
			RANGE(even, 0, 100);
			break;

		case 'f':
			param.freq = atoi(optarg);
			RANGE(freq, 50, 4000);
			break;

		case 'n':
			param.number = atoi(optarg);
			RANGE(number, 1, 100);
			break;

		case 'h':
			param.hum = atoi(optarg);
			RANGE(hum, 0, 100);
			break;

		case 'H':
			param.hand = atoi(optarg);
			RANGE(hand, 0, 100);
			break;

		case 'l':
			param.click = atoi(optarg);
			RANGE(click, 0, 100);
			break;

		case 'L':
			param.spacelen = atoi(optarg);
			if(param.spacelen == 0) param.spacelen = 100;
			RANGE(spacelen, 20, 300);
			break;

		case 'N':
			param.noise = atoi(optarg);
			RANGE(noise, 0, 100);
			break;

		case 'o':
			strncpy(filename, optarg, 256);
			break;

		case 'O':
			param.odd = atoi(optarg);
			RANGE(odd, 0, 100);
			break;

		case 'p':
			param.sweep = atoi(optarg);
			RANGE(sweep, -4000, 4000);
			break;

		case 'P':
			param.sweepness = atoi(optarg);
			RANGE(sweepness, 0, 10000);
			break;

		case 'q':
			param.qsb = atoi(optarg);
			RANGE(qsb, 0, 100);
			break;

		case 'r':
			if((val = atoi(optarg))) param.seed = val;
			RANGE(seed, 0, 32767);
			break;

		case 's':
			param.cspaces = atoi(optarg);
			RANGE(cspaces, 0, 100);
			break;

		case 'S':
			param.signals = atoi(optarg);
			RANGE(signals, 1, 5);
			break;

		case 't':
			param.tempo = atoi(optarg);
			RANGE(tempo, 5, 500);
			break;

		case 'v':
			wordset = atoi(optarg);
			if(wordset < 10) wordset = 10;
			if(wordset > 1000) wordset = 1000;
			break;

		case 'w':
			param.window = atoi(optarg);
			RANGE(window, 0, 1000);
			break;

		case 'W':
			param.wspaces = atoi(optarg);
			RANGE(wspaces, 0, 100);
			break;

		case 'x':
			param.lowcut = atoi(optarg);
			RANGE(lowcut, 50, 1000);
			break;

		case 'X':
			param.shape = atoi(optarg);
			RANGE(shape, -50, 50);
			break;

		case 'y':
			param.highcut = atoi(optarg);
			RANGE(highcut, 300, 10000);
			break;
		}
	}

	if(output) {
		fprintf(stderr, "\n----------------------------------------------------\n");
		fprintf(stderr, "CWStudio %s\nCopyright 2009-2011 Lukasz Komsta, SP8QED\n", VERSION);
		fprintf(stderr, "Licensed under GPLv3\n");
		fprintf(stderr, "----------------------------------------------------\n");
		fprintf(stderr, "* Working at %i Hz, %i bits\n", samplerate, bits);
	}

	/* If stdin is injected via pipe, read it. Otherwise, generate random group */
	if(isatty(STDIN_FILENO)) {
		switch(mode)
		{
		case 0:
			if((text = cw_rand_groups(param.number, param.shape, charset, param.seed)) == NULL) return(CWALLOC);
			if(output) fprintf(stderr, "* Charset: %s\n\n", charset);
			break;

		case 1:
			if((text = cw_rand_words(param.number, param.shape, wordset, param.seed)) == NULL) return(CWALLOC);
			if(output) fprintf(stderr, "* %i words from %i most common\n\n", param.number, wordset);
			break;

		case 2:
			if((text = cw_rand_calls(param.number, param.shape, param.seed)) == NULL) return(CWALLOC);
			if(output) fprintf(stderr, "* %i calls\n\n", param.number);
			break;
		}
	}
	else {
		fprintf(stderr, "* Getting text from stdin\n\n");
		if((text = malloc(2048 * sizeof(char))) == NULL) return(CWALLOC);
		(void) fgets(text, 2048, stdin);
	}

	/* Encode text */
	if((morsetext = cw_encode(text)) == NULL) return(CWALLOC);

	/* Next part of output */
	if(output) {
		fprintf(stderr, "* Frequency %i Hz * Window %i samples\n", param.freq, param.window);
		if(param.even) fprintf(stderr, "* %i%% even harmonics ", param.even);
		if(param.odd) fprintf(stderr, "* %i%% odd harmonics ", param.odd);
		if(param.even || param.odd) fprintf(stderr, "\n");
		if(param.click) fprintf(stderr, "* %i dB Click ", param.click);
		if(param.hand) fprintf(stderr, "* Hand simulation %i%% ", param.detune);
		if(param.hum) fprintf(stderr, "* %i%% Hum ", param.hum);
		fprintf(stderr, "\n");
		if(param.sweepness) fprintf(stderr, "* Sweep from %i Hz, sweepness %i\n", param.sweep, param.sweepness);
		if(param.detune) fprintf(stderr, "* Detune %i%% ", param.detune);
		if(param.qsb) fprintf(stderr, "* QSB %i%% ", param.detune);
		fprintf(stderr, "\n");
		fprintf(stderr, "* Tempo is %i wpm ", param.tempo);
		if(param.cspaces) fprintf(stderr, "* Char spacing +%i ", param.cspaces);
		if(param.wspaces) fprintf(stderr, "* Word spacing +%i ", param.wspaces);
		fprintf(stderr, "\n");
		fprintf(stderr, "* Random seed: %i ", param.seed);
		if(param.shape)
			fprintf(stderr, "* Random shape: %i \n", param.shape);
		else
			fprintf(stderr, "\n");
		if(param.signals > 1) fprintf(stderr, "* Mixing %i signals ", param.signals);
		fprintf(stderr, "\n");
		if(param.noise) {
			fprintf(stderr, "* Adding %i%% noise, %i - %i Hz ", param.noise, param.lowcut, param.highcut);
			if(param.agc) fprintf(stderr, "* %i%% AGC ", param.agc);
			fprintf(stderr, "\n");
		}

		if(param.dashlen != 300) fprintf(stderr, "* Dash length: %i%% ", param.dashlen);
		if(param.spacelen != 100) fprintf(stderr, "* Space length: %i%% ", param.spacelen);
		if((param.dashlen != 300) || (param.spacelen != 100)) fprintf(stderr, "\n");
		fprintf(stderr, "----------------------------------------------------\n\n");
	}

	fprintf(stderr, "\n%s\n\n", text);

	/*
	 * "asound" is floating sample created by library, converted sample goest to
	 * "csound"
	 */
	cw_initsample(&asound, NULL);
	asound.samplerate = samplerate;
	cw_initsample(&csound, &asound);

	/* Sound generation */
	if((err = cw_signals(&asound, param, morsetext)) != CWOK) return(err);
	if((err = cw_convert(&asound, &csound, bits)) != CWOK) return(err);

	/* If stdout is redirected somewhere, feed generated WAV file there. */
	if(isatty(STDOUT_FILENO)) {
		if((i = cw_wavout(filename, &csound)) != CWOK) return(i);
	}
	else {
		fprintf(stderr, "* Redirecting sound to stdout\n\n");
		if((i = cw_wavout(NULL, &csound)) != CWOK) return(i);
	}

	fflush(stderr);

	/* Play if needed */
	if(play) {
		fprintf(stderr, "Playing...");
		fflush(stderr);
		playsample(&csound);
		fprintf(stderr, "\n\n");
	}

	/* Free memory */
	cw_freesample(&asound);
	cw_freesample(&csound);
	cw_free(text);
	cw_free(morsetext);

	/* Exit */
	return(CWOK);
}
